File: SpaceInvaders/GameObject/GameObject.cs
using System;
using System.Diagnostics;

namespace SpaceInvaders
{
    public abstract class GameObject : CollisionVisitor
    {
        protected GameObject.Name name;
        protected bool markedForDeath;
        protected float x;
        protected float y;
        protected SpriteProxy pSpriteProxy;
        protected CollisionObject pCollisionObject;

        //----------------------------------------------------------------------
        // Constructors / Destructor
        //----------------------------------------------------------------------

        protected GameObject(GameObject.Name gameName, Sprite.Name spriteName)
        {
            this.name = gameName;
            this.x = 0.0f;
            this.y = 0.0f;
            this.pSpriteProxy = new SpriteProxy(spriteName);
            this.pCollisionObject = new CollisionObject(this.pSpriteProxy);
        }

        ~GameObject()
        {
            this.name = GameObject.Name.Uninitialized;
            this.pSpriteProxy = null;
            this.pCollisionObject = null;
        }

        //----------------------------------------------------------------------
        // Methods
        //----------------------------------------------------------------------
        public virtual void Update()
        {
            Debug.Assert(this.pSpriteProxy != null);
            this.pSpriteProxy.SetX(this.x);
            this.pSpriteProxy.SetY(this.y);

            Debug.Assert(this.pCollisionObject != null);
            this.pCollisionObject.UpdatePos(this.x, this.y);

            SpriteBox pCollisionSpriteBox = this.pCollisionObject.GetCollisionSpriteBox();
            Debug.Assert(pCollisionSpriteBox != null);
            pCollisionSpriteBox.Update();
        }

        protected void BaseUpdateBoundingBox()
        {
            // Go to first child
            Component pComponent = (Component)this;
            GameObject pNode = (GameObject)ForwardIterator.GetChild(pComponent);
            Debug.Assert(pNode != null);

            CollisionRect ColTotal = this.pCollisionObject.GetCollisionRect();
            ColTotal.Set(pNode.GetCollisionObject().GetCollisionRect());

            // loop through sliblings
            while (pNode != null)
            {
                ColTotal.Union(pNode.GetCollisionObject().GetCollisionRect());

                // go to next sibling
                pComponent = ForwardIterator.GetSibling(pNode);

                if (pComponent != null)
                {
                    pNode = (GameObject)pComponent;
                }
                else
                {
                    pNode = null;
                }
            }

            this.x = this.GetCollisionObject().GetCollisionRect().x;
            this.y = this.GetCollisionObject().GetCollisionRect().y;

            //Debug.WriteLine("x:{0} y:{1} w:{2} h:{3}", ColTotal.x, ColTotal.y, ColTotal.width, ColTotal.height);
        }

        public void ActivateCollisionSprite(SpriteBatch pSpriteBatch)
        {
            Debug.Assert(pSpriteBatch != null);
            Debug.Assert(this.pCollisionObject != null);
            pSpriteBatch.Attach(this.pCollisionObject.GetCollisionSpriteBox());
        }
        public void ActivateSprite(SpriteBatch pSpriteBatch)
        {
            Debug.Assert(pSpriteBatch != null);
            pSpriteBatch.Attach(this.pSpriteProxy);
        }

        public GameObject.Name GetName()
        {
            return this.name;
        }

        public void SetName(Name newName)
        {
            this.name = newName;
        }

        public float GetX()
        {
            return this.x;
        }

        public void SetX(float x)
        {
            this.x = x;
        }

        public float GetY()
        {
            return this.y;
        }

        public void SetY(float y)
        {
            this.y = y;
        }

        public SpriteProxy GetSpriteProxy()
        {
            return this.pSpriteProxy;
        }

        public CollisionObject GetCollisionObject()
        {
            return this.pCollisionObject;
        }

        public void SetPos(float inX, float inY)
        {
            this.x = inX;
            this.y = inY;
        }

        public void SetCollisionObjectLineColor(float r, float g, float b)
        {
            this.GetCollisionObject().GetCollisionSpriteBox().SetLineColor(r, g, b);
        }

        public bool IsMarkedForDeath()
        {
            return markedForDeath;
        }

        public void MarkForDeath()
        {
            markedForDeath = true;
        }

        public virtual void Remove(SpriteBatch.Name gameObjSpriteBatch, SpriteBatch.Name collisionBoxSpriteBatch)
        {
            // Very difficult at first... if you are messy, you will pay here!
            // Given a game object....

            // Since the Root object is being drawn
            // 1st set its size to zero
            this.pCollisionObject.GetCollisionRect().Set(0, 0, 0, 0);
            this.Update();

            // Update the parent
            GameObject pParent = (GameObject)this.pParent;
            if(pParent != null)
            {
                pParent.Update();
            }

            // Remove from SpriteBatch
            GameStateManager.GetGame().GetStateSpriteBatchManager().Find(gameObjSpriteBatch).Detach(pSpriteProxy);
            GameStateManager.GetGame().GetStateSpriteBatchManager().Find(collisionBoxSpriteBatch).Detach(pCollisionObject.GetCollisionSpriteBox());

            // Remove from GameObjectManager
            GameStateManager.GetGame().GetStateGameObjectManager().Detach(this);

            // Attach to GhostManager for later use
            GameStateManager.GetGame().GetStateGhostManager().Attach(this);

            // Reset death flag
            this.markedForDeath = false;

        }

        public enum Name
        {
            SquidAlien,
            JellyFishAlien,
            CrabAlien,
            FlyingSaucer,
            CoreCannon,

            AlienGrid,
            AlienGridColumn,

            Missile,
            MissileGroup,

            
            ShieldGroup,
            ShieldRoot,
            ShieldBrick,
            ShieldColumn,

            RightWall,
            LeftWall,
            TopWall,
            BottomWall,

            CoreCannonGroup,
            
            BombRoot,
            Bomb,

            Null_Object,
            Uninitialized,
            LeftBumper,
            RightBumper,
            FlyingSaucerRoot,
            FlyingSaucerBomb,
            ShieldBrick_LeftTop1,
            ShieldBrick_LeftTop0,
            ShieldBrick_LeftBottom,
            ShieldBrick_RightBottom,
            ShieldBrick_RightTop1,
            ShieldBrick_RightTop0,
        }

    }

}

File: quarkus-migration/space-invaders-quarkus/src/main/java/net/beeland/spaceinvaders/gameobject/GameObject.java
package net.beeland.spaceinvaders.gameobject;

import net.beeland.spaceinvaders.collision.CollisionObject;
import net.beeland.spaceinvaders.collision.CollisionRect;
import net.beeland.spaceinvaders.composite.Component;
import net.beeland.spaceinvaders.sprite.SpriteProxy;

/**
 * GameObject is the abstract base class for all game entities.
 * It combines the Composite pattern (via Component) with collision detection
 * and sprite rendering capabilities.
 * 
 * Each GameObject has:
 * - A name for identification
 * - Position (x, y coordinates)
 * - A sprite proxy for rendering
 * - A collision object for collision detection
 * - A death flag for lifecycle management
 */
public abstract class GameObject extends Component {
    
    protected GameObjectName name;
    protected boolean markedForDeath;
    protected float x;
    protected float y;
    protected SpriteProxy spriteProxy;
    protected CollisionObject collisionObject;

    /**
     * Creates a game object with the specified name.
     * Sprite proxy must be set separately.
     *
     * @param gameName The name identifier for this game object
     */
    protected GameObject(GameObjectName gameName) {
        this.name = gameName;
        this.x = 0.0f;
        this.y = 0.0f;
        this.markedForDeath = false;
        this.spriteProxy = new SpriteProxy();
        this.collisionObject = new CollisionObject(this.spriteProxy);
    }

    /**
     * Updates the game object's state.
     * This includes updating sprite position and collision bounds.
     */
    public void update() {
        if (this.spriteProxy != null) {
            this.spriteProxy.setPosition(this.x, this.y);
        }

        if (this.collisionObject != null) {
            this.collisionObject.updatePos(this.x, this.y);
        }
    }

    /**
     * Gets the name of this game object.
     *
     * @return The game object name
     */
    public GameObjectName getName() {
        return this.name;
    }

    /**
     * Sets the name of this game object.
     *
     * @param name The new name
     */
    public void setName(GameObjectName name) {
        this.name = name;
    }

    /**
     * Gets the X coordinate.
     *
     * @return The X coordinate
     */
    public float getX() {
        return this.x;
    }

    /**
     * Sets the X coordinate.
     *
     * @param x The new X coordinate
     */
    public void setX(float x) {
        this.x = x;
    }

    /**
     * Gets the Y coordinate.
     *
     * @return The Y coordinate
     */
    public float getY() {
        return this.y;
    }

    /**
     * Sets the Y coordinate.
     *
     * @param y The new Y coordinate
     */
    public void setY(float y) {
        this.y = y;
    }

    /**
     * Sets the position of this game object.
     *
     * @param x The X coordinate
     * @param y The Y coordinate
     */
    public void setPos(float x, float y) {
        this.x = x;
        this.y = y;
    }

    /**
     * Gets the sprite proxy for this game object.
     *
     * @return The sprite proxy
     */
    public SpriteProxy getSpriteProxy() {
        return this.spriteProxy;
    }

    /**
     * Gets the collision object for this game object.
     *
     * @return The collision object
     */
    public CollisionObject getCollisionObject() {
        return this.collisionObject;
    }

    /**
     * Checks if this game object is marked for death.
     *
     * @return true if marked for death, false otherwise
     */
    public boolean isMarkedForDeath() {
        return this.markedForDeath;
    }

    /**
     * Marks this game object for death (to be removed in next update cycle).
     */
    public void markForDeath() {
        this.markedForDeath = true;
    }

    /**
     * Clears the death flag.
     */
    public void clearDeathFlag() {
        this.markedForDeath = false;
    }

    /**
     * Accepts a collision visitor for double-dispatch collision handling.
     * Subclasses must implement this to define collision behavior.
     *
     * @param other The collision visitor
     */
    public abstract void accept(CollisionVisitor other);

    /**
     * Visits another game object for collision handling.
     * Subclasses can override to define specific collision responses.
     *
     * @param other The game object being visited
     */
    public void visit(GameObject other) {
        // Default implementation - can be overridden by subclasses
    }

    /**
     * Updates the bounding box for composite game objects.
     * This method calculates the union of all child collision rectangles.
     */
    protected void baseUpdateBoundingBox() {
        // Get first child
        Component component = this;
        GameObject node = (GameObject) getChild(component);
        
        if (node == null) {
            return;
        }

        CollisionRect totalRect = this.collisionObject.getCollisionRect();
        totalRect.set(node.getCollisionObject().getCollisionRect());

        // Loop through siblings
        while (node != null) {
            totalRect.union(node.getCollisionObject().getCollisionRect());

            // Go to next sibling
            component = getSibling(node);
            node = (component != null) ? (GameObject) component : null;
        }

        this.x = this.collisionObject.getCollisionRect().x;
        this.y = this.collisionObject.getCollisionRect().y;
    }

    /**
     * Helper method to get the first child of a component.
     *
     * @param component The parent component
     * @return The first child, or null if none exists
     */
    private Component getChild(Component component) {
        return component.getFirstChild();
    }

    /**
     * Helper method to get the next sibling of a component.
     *
     * @param component The current component
     * @return The next sibling, or null if none exists
     */
    private Component getSibling(Component component) {
        return (Component) component.getNext();
    }

    /**
     * Washes/resets this game object for reuse in the object pool.
     */
    @Override
    protected void wash() {
        this.name = GameObjectName.UNINITIALIZED;
        this.x = 0.0f;
        this.y = 0.0f;
        this.markedForDeath = false;
        if (this.spriteProxy != null) {
            this.spriteProxy.wash();
        }
    }

    /**
     * Dumps debug information about this game object.
     */
    @Override
    public void dump() {
        System.out.println(toString());
    }

    @Override
    public String toString() {
        return String.format("%s[name=%s, pos=(%.2f, %.2f), dead=%b]",
                getClass().getSimpleName(), name, x, y, markedForDeath);
    }

    /**
     * Enum defining all possible game object types.
     */
    public enum GameObjectName {
        // Aliens
        SQUID_ALIEN,
        JELLYFISH_ALIEN,
        CRAB_ALIEN,
        FLYING_SAUCER,
        FLYING_SAUCER_ROOT,
        
        // Player
        CORE_CANNON,
        CORE_CANNON_GROUP,
        
        // Alien structures
        ALIEN_GRID,
        ALIEN_GRID_COLUMN,
        
        // Projectiles
        MISSILE,
        MISSILE_GROUP,
        BOMB,
        BOMB_ROOT,
        FLYING_SAUCER_BOMB,
        
        // Shields
        SHIELD_GROUP,
        SHIELD_ROOT,
        SHIELD_BRICK,
        SHIELD_COLUMN,
        SHIELD_BRICK_LEFT_TOP_0,
        SHIELD_BRICK_LEFT_TOP_1,
        SHIELD_BRICK_LEFT_BOTTOM,
        SHIELD_BRICK_RIGHT_BOTTOM,
        SHIELD_BRICK_RIGHT_TOP_0,
        SHIELD_BRICK_RIGHT_TOP_1,
        
        // Walls
        RIGHT_WALL,
        LEFT_WALL,
        TOP_WALL,
        BOTTOM_WALL,
        LEFT_BUMPER,
        RIGHT_BUMPER,
        
        // Special
        NULL_OBJECT,
        UNINITIALIZED
    }

    /**
     * Interface for collision visitor pattern.
     * Allows double-dispatch for collision handling.
     */
    public interface CollisionVisitor {
        void visit(GameObject other);
        
        default void visitAlienGrid(AlienGrid alienGrid) {
            // Default implementation - can be overridden
        }
        
        default void visitAlienGridColumn(AlienGridColumn alienGridColumn) {
            // Default implementation - can be overridden
        }
        
        default void visitMissileGroup(MissileGroup missileGroup) {
            // Default implementation - can be overridden
        }
        
        default void visitCoreCannonGroup(CoreCannonGroup coreCannonGroup) {
            // Default implementation - can be overridden
        }
        
        default void visitShieldBrick(ShieldBrick shieldBrick) {
            // Default implementation - can be overridden
        }
        
        default void visitFlyingSaucer(FlyingSaucer flyingSaucer) {
            // Default implementation - can be overridden
        }
        
        default void visitFlyingSaucerRoot(net.beeland.spaceinvaders.composite.Composite flyingSaucerRoot) {
            // Default implementation - can be overridden
        }
    }
}